File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Session/OriginState/DatabaseOriginState.cs
// <copyright file="DatabaseOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using Allors.Protocol.Json.Api.Push;

    internal sealed class DatabaseOriginState : Adapters.DatabaseOriginState
    {
        internal DatabaseOriginState(Strategy strategy, DatabaseRecord record) : base(record) => this.RemoteStrategy = strategy;

        public override Adapters.Strategy Strategy => this.RemoteStrategy;
        private Strategy RemoteStrategy { get; }

        internal PushRequestNewObject PushNew() => new PushRequestNewObject
        {
            WorkspaceId = this.Id,
            ObjectType = this.Class.Tag,
            Roles = this.PushRoles()
        };

        internal PushRequestObject PushExisting() => new PushRequestObject
        {
            DatabaseId = this.Id,
            Version = this.Version,
            Roles = this.PushRoles()
        };

        private PushRequestRole[] PushRoles()
        {
            if (this.ChangedRoleByRelationType?.Count > 0)
            {
                var numbers = this.RemoteStrategy.Session.Workspace.Database.Numbers;

                var roles = new List<PushRequestRole>();

                foreach (var keyValuePair in this.ChangedRoleByRelationType)
                {
                    var relationType = keyValuePair.Key;
                    var roleValue = keyValuePair.Value;

                    var pushRequestRole = new PushRequestRole { RelationType = relationType.Tag };

                    if (relationType.RoleType.ObjectType.IsUnit)
                    {
                        pushRequestRole.SetUnitRole = UnitConvert.ToJson(roleValue);
                    }
                    else
                    {
                        if (relationType.RoleType.IsOne)
                        {
                            pushRequestRole.SetCompositeRole = (long?)roleValue;
                        }
                        else
                        {
                            if (!this.ExistDatabaseRecord)
                            {
                                pushRequestRole.AddCompositesRole = numbers.ToArray(roleValue);
                            }
                            else
                            {
                                var databaseRole = (long[])this.DatabaseRecord.GetRole(relationType.RoleType);
                                if (databaseRole == null)
                                {
                                    pushRequestRole.AddCompositesRole = numbers.ToArray(roleValue);
                                }
                                else
                                {
                                    pushRequestRole.AddCompositesRole = numbers.ToArray(numbers.Except(roleValue, databaseRole));
                                    pushRequestRole.RemoveCompositesRole = numbers.ToArray(numbers.Except(databaseRole, roleValue));
                                }
                            }
                        }
                    }

                    roles.Add(pushRequestRole);
                }

                return roles.ToArray();
            }

            return null;
        }
    }
}

File: typescript/libs/workspace/adapters/json/system/src/lib/runtime/DatabaseOriginState.ts
import { PushRequestNewObject, PushRequestObject, PushRequestRole } from '@allors/protocol/json/system';
import { DatabaseOriginState as SystemDatabaseOriginState, DatabaseRecord, difference, Strategy, save, IRange } from '@allors/workspace/adapters/system';
import { unitToJson } from '../json/toJson';

export class DatabaseOriginState extends SystemDatabaseOriginState {
  constructor(strategy: Strategy, record: DatabaseRecord) {
    super(strategy, record);
  }

  PushNew(): PushRequestNewObject {
    return {
      w: this.id,
      t: this.class.tag,
      r: this.PushRoles(),
    };
  }

  PushExisting(): PushRequestObject {
    return {
      d: this.id,
      v: this.version,
      r: this.PushRoles(),
    };
  }

  private PushRoles(): PushRequestRole[] {
    if (this.changedRoleByRelationType?.size > 0) {
      const roles: PushRequestRole[] = [];

      for (const [relationType, roleValue] of this.changedRoleByRelationType) {
        const pushRequestRole: PushRequestRole = { t: relationType.tag };

        if (relationType.roleType.objectType.isUnit) {
          pushRequestRole.u = unitToJson(roleValue);
        } else if (relationType.roleType.isOne) {
          pushRequestRole.c = roleValue as number;
        } else {
          const roleIds = roleValue as IRange;
          if (!this.existRecord) {
            pushRequestRole.a = save(roleIds);
          } else {
            const databaseRole = this.databaseRecord.getRole(relationType.roleType) as IRange;
            if (databaseRole == null) {
              pushRequestRole.a = save(roleIds);
            } else {
              pushRequestRole.a = save(difference(roleIds, databaseRole));
              pushRequestRole.r = save(difference(databaseRole, roleIds));
            }
          }
        }

        roles.push(pushRequestRole);
      }

      return roles;
    }

    return null;
  }
}
